core/src/main/java/com/datastax/oss/driver/api/core/type/codec/MappingCodec.java

```diff
@@ -0,0 +1,156 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core.type.codec;
+
+import com.datastax.oss.driver.api.core.ProtocolVersion;
+import com.datastax.oss.driver.api.core.type.DataType;
+import com.datastax.oss.driver.api.core.type.reflect.GenericType;
+import com.datastax.oss.driver.shaded.guava.common.base.Preconditions;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+/**
+ * A {@link TypeCodec} that maps instances of {@code InnerT}, a supported Java type, to instances of
+ * a target {@code OuterT} Java type.
+ *
+ * <p>This codec can be used to provide support for Java types that are not natively handled by the
+ * driver, as long as there is a conversion path to and from another supported Java type.
+ *
+ * <p>See the <a
+ * href="https://docs.datastax.com/en/developer/java-driver/latest/manual/core/custom_codecs/">driver
+ * documentation on custom codecs</a> for more information and a few usage examples.
+ *
+ * @param <InnerT> The "inner" Java type; must be a driver supported Java type (that is, there must
+ *     exist a codec registered for it).
+ * @param <OuterT> The "outer", or target Java type; this codec will handle the mapping to and from
+ *     {@code InnerT} and {@code OuterT}.
+ */
+public abstract class MappingCodec<InnerT, OuterT> implements TypeCodec<OuterT> {
+
+  protected final TypeCodec<InnerT> innerCodec;
+  protected final GenericType<OuterT> outerJavaType;
+
+  /**
+   * Creates a new mapping codec providing support for {@code OuterT} based on an existing codec for
+   * {@code InnerT}.
+   *
+   * @param innerCodec The inner codec to use to handle instances of InnerT; must not be null.
+   * @param outerJavaType The outer Java type; must not be null.
+   */
+  public MappingCodec(
+      @NonNull TypeCodec<InnerT> innerCodec, @NonNull GenericType<OuterT> outerJavaType) {
+    this.innerCodec = Objects.requireNonNull(innerCodec, "innerCodec cannot be null");
+    this.outerJavaType = Objects.requireNonNull(outerJavaType, "outerJavaType cannot be null");
+  }
+
+  /** @return The type of {@code OuterT}. */
+  @NonNull
+  @Override
+  public GenericType<OuterT> getJavaType() {
+    return outerJavaType;
+  }
+
+  /** @return The type of {@code InnerT}. */
+  public GenericType<InnerT> getInnerJavaType() {
+    return innerCodec.getJavaType();
+  }
+
+  /**
+   * Whether this codec is capable of encoding the given Java object.
+   *
+   * <p>This implementation first checks if the value's runtime type is compatible with {@code
+   * OuterT}, then checks if the converted value's runtime type is compatible with {@code InnerT}.
```

Comment:
I think it is, at least for correctness. With the example of `CqlIntToStringCodec` in mind, without this extra check the method would return true for the String `"NaN"`, but that string cannot be encoded as a CQL int.

That said, I agree that if we remove the extra check, there would probably be no visible consequence: let's assume that we are calling `CodecRegistry.codecFor(DataTypes.INT, "123")`. The built-in codec for `DataTypes.INT` does not accept `"123"` and would not be selected. Only a user codec accepting both `DataTypes.INT` and an instance of String can be selected, and so `CqlIntToStringCodec` is the only codec that could be returned, regardless if the extra check is present or not.

Reply:
I'm revisiting this before the merge and I think I agree we can avoid the extra conversion. The concrete consequence is that `codec.accepts("not an int")` will return true for a `MappingCodec<Integer,String>`. I'm going to merge this because I need it for JAVA-2691, we can discuss this offline if you wish.